### ISSUE TITLE
`zigHook` probably shouldn't rely on `-Dcpu=baseline`

### ISSUE THREAD

**Issue by sagehane** [NixOS/nixpkgs#247719]
### Describe the bug
https://github.com/NixOS/nixpkgs/pull/242397 introduced the `zigHook`.

Currently, it aims for reproducibility by passing the `-Dcpu=baseline` flag, as shown here:
https://github.com/NixOS/nixpkgs/blob/a85a699e40708b452d36c70725396ff67dc48e66/pkgs/development/compilers/zig/setup-hook.sh#L23

With the current `zigHook`, I have two concerns:
1. Is it possible to cross-compile Zig packages using Nix?

Cross-compiling `poop`:
```sh-session
$ git describe
22.05-pre-177775-ga157ed02d34
$ nix build .#pkgsCross.aarch64-multiplatform.poop
error: builder for '/nix/store/9zzp86jir6bc64vv3b9h7i69ams4kaix-poop-aarch64-unknown-linux-gnu-0.4.0.drv' failed with exit code 1;
       last 10 log lines:
       > Hunk #6 succeeded at 530 (offset -8 lines).
       > Hunk #7 succeeded at 647 (offset -8 lines).
       > patching file src/progress.zig
       > updateAutotoolsGnuConfigScriptsPhase
       > updateAutotoolsGnuConfigScriptsPhase
       > configuring
       > no configure script, doing nothing
       > building
       > build flags: -Doptimize=ReleaseSafe -Dcpu=baseline
       > error: AccessDenied
       For full logs, run 'nix log /nix/store/9zzp86jir6bc64vv3b9h7i69ams4kaix-poop-aarch64-unknown-linux-gnu-0.4.0.drv'.
```
I chose `poop` as it's a small and pure Zig project maintained by one of the Zig members. `waylock` might also make a good candidate for a package that makes use of C interop. Not too sure if this build issue is caused by the current behaviour of `zigHook`, however.

2. Can users pass arbitrary targets to Zig packages built with Nix, or would the default flag cause a compile error?

Example of passing a new baseline to a build system on top of an existing one:
```sh-session
$ zig init-exe
info: Created build.zig
info: Created src/main.zig
info: Next, try `zig build --help` or `zig build run`
$ zig build -Dcpu=baseline -Dcpu=x86_64_v4
error: Expected -Dcpu to be a string, but received a list.
<omitted>
```

### Expected behavior
I'll quote @andrewrk's comment at https://github.com/NixOS/nixpkgs/issues/185665#issuecomment-1416592467:
>I think a better way would be accessing the desired target architecture, operating system along with minimum/maximum OS version, and ABI, along with potentially the glibc version, from nix, and then specifying these via the standard `-Dtarget` option. I'm happy to assist in getting this working.

### Additional context
I think the following Guix patch by ekaitz is a great example: https://issues.guix.gnu.org/60889.

In the future, https://github.com/ziglang/zig/issues/14281#issuecomment-1624220653 should be implemented so one can pass `--override-host-target <target>` and `--override-host-cpu <cpu>` instead.

### Notify maintainers
@AndersonTorres for writing the hook?

**Comment by AndersonTorres** [comment]
As far as I remember, many of those things were documented on the huge comment I wrote - including the links you are citing.

What is exactly the matter?

**Comment by sagehane** [comment]
Ah, sorry, I must have missed the comment here:
https://github.com/NixOS/nixpkgs/blob/6acf0bf645f8826940bc7aca9efd46fbdae69eae/pkgs/development/compilers/zig/setup-hook.sh#L3-L21

>Hopefully they [the necessary features] will be implemented in future releases of Zig. When this happens, this flagset should be revisited accordingly.

With this in mind, I guess my issue is kind of meaningless because it's true that the code will have to change in the foreseeable future and people can worry about it then.

But sure, let me rephrase what I was trying to say. I think it would still be more preferable to properly pass a target triple than relying on `-Dcpu=baseline`. I'm not quite sure how cross-compilation works on Nix (still trying to figure out https://nixos.org/manual/nixpkgs/stable/#chap-cross), but wouldn't `-Dcpu=baseline` cross-compile use the `buildPlatform` instead of the `hostPlatform` as the target and thus always fail?

**Comment by AndersonTorres** [comment]
Indeed I need to look at this cross-compilation issue too, because I don't remember if there is an explicit setting like `-Dhost-cpu` and `-Dtarget-cpu`.

There is the issue of reproducibility too. This set of default flags fixed some nasty miscompilations, as explained in the links.

**Comment by sagehane** [comment]
Assuming that cross-compilation is broken (like I said, I'm not completely sure but it seems like it?), the way to fix it would be make sure the correct target is provided such that compiling to Aarch64 Linux would make the hook pass `-Dtarget=aarch64-linux`. Always specifying the correct `-Dtarget` value would remove the assumption that the host is always the same as build platform.

Guix's `zig-build-system` (mentioned above) does this and I think it looks great.

And if one wants to be more specific (say, targeting a dev board using A53 cores), one can just do `-Dtarget=aarch64-linux -Dcpu=cortex_a53`.

Also, just found https://github.com/flyx/Zicross and they also seem to use `-Dtarget` for cross compilation.

**Comment by Cloudef** [comment]
Zig author recommends passing `-target` over `-mcpu=baseline` here as well
https://ziggit.dev/t/exe-files-not-interchangeable-among-identical-linux-systems/2708/6?u=cloudef

**Comment by Cloudef** [comment]
I ended up replacing nixpkgs hook here https://github.com/Cloudef/zig2nix/commit/fd7931adc3e12e5458204d28ece7b2d38173c0a0 
The default flags are also problematic as you can't override them. For zig it seems to be a error to pass multiple -Doptimize so you can't build package with a different release mode for example.

**Comment by AndersonTorres** [comment]
Still broken?
